src/librustc_mir/interpret/eval_context.rs

```diff
@@ -33,11 +33,8 @@ pub struct EvalContext<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>> {
     /// Bounds in scope for polymorphic evaluations.
     pub param_env: ty::ParamEnv<'tcx>,
 
-    /// The virtual memory system.
```

Comment:
The `machine` field also needs to go into the `EvalState`. It's unused in the compiler, but `miri` the tool needs it

Reply:
Got it, when I first started on this, I assumed that I'd need to provide some custom implementations of common methods on `EvalState`, as well as pass it by reference to something like `HashSet::get`. However, since I implemented custom Eq/Hash on `Frame` and `Memory`, `observe` could simply accept several references to individulal fields.

This means I could revert my first commit and not break everything.

Reply:
Also, I could use #[derive] more as well as remove a lot of where bounds if I add `Eq + Hash + Clone` to the supertraits of `Machine`. What's rustc's general philosophy on this? I try to avoid needlessly adding supertraits, but here it would save a lot of boilerplate.